### PR TITLE
Add RelocateAlternative

### PR DESCRIPTION
**Pull request by leonlan**
This PR adds RelocateAlternative. This is useful for VRPs with mutually exclusive groups, and covers a part of the search space that wasn't covered before. We noticed it when benchmarking #1073; our benchmark results show a fairly big decrease in average gap:

| Experiment | Mean gap |
|---|---:|
| Main | 3.135% |
| RelocateAlternative | 2.051% |

The same holds for GVRP.

| Experiment | Mean gap |
|---|---:|
| Main | 1.064% |
| RelocateAlternative | 0.715% |

### CODE REVIEW

File: benchmarks/conftest.py
Comment (leonlan) [inline benchmarks/conftest.py:1]: Adding this benchmark covers two blind spots: mutually exclusive groups and TSP.

File: pyvrp/cpp/search/RelocateWithAlternative.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+std::string RelocateWithAlternative::name() const
+{
+    return "RelocateWithAlternative";
+}
+
+template <>
+bool pyvrp::search::supports<RelocateWithAlternative>(ProblemData const &data)
+{
+    return data.numGroups() > 0;
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateWithAlternative.cpp]: This should check whether any of the groups are mutually exclusive. We do not currently have any other type of group, but I did plan at some point for that to change.

Something like
```cpp
for (auto const &group : data.groups())
  if (group.mutuallyExclusive)
    return true;

return false;
```

File: pyvrp/cpp/search/RelocateWithAlternative.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    auto const *vRoute = V->route();
+    auto const &uData = data.client(U->idx());
+    if (!uRoute || !vRoute || !uData.group || U == V || U == n(V))
+        return std::make_pair(0, false);
+
+    if (uRoute == vRoute && U->trip() != V->trip())
+        return std::make_pair(0, false);
+
+    auto const &group = data.group(*uData.group);
+    assert(group.mutuallyExclusive);
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateWithAlternative.cpp]: I think we're wrong in asserting this if we ever change the group system, but we do so in other places too, so should be OK for now.

File: pyvrp/cpp/search/RelocateWithAlternative.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        else
+        {
+            costEvaluator.deltaCost(
+                deltaCost,
+                Route::Proposal(uRoute->before(V->pos()),
+                                ClientSegment(data, client),
+                                uRoute->between(V->pos() + 1, U->pos() - 1),
+                                uRoute->after(U->pos() + 1)));
+        }
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateWithAlternative.cpp]: Move if/else outside the loop, it does not depend on anything looping related

File: pyvrp/cpp/search/RelocateWithAlternative.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        if (deltaCost < bestCost)
+        {
+            bestCost = deltaCost;
+            alternative_ = alternative;
+        }
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateWithAlternative.cpp]: Do we need to do best improvement? This now exhausts groups for each evaluation, which can be expensive for big groups. Maybe first improvement?

File: pyvrp/cpp/search/RelocateWithAlternative.cpp (hunk 3 above)
Comment (N-Wouda) [reply pyvrp/cpp/search/RelocateWithAlternative.cpp]: then loop inside each branch (maybe separate member functions if necessary, this is all rather big already)

File: pyvrp/cpp/search/RelocateWithAlternative.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace pyvrp::search
+{
+/**
+ * RelocateWithAlternative(data: ProblemData)
+ *
+ * Evaluates replacing the current mutually exclusive group member :math:`U` by
+ * another member of the same group, while relocating it after :math:`V`.
+ */
+class RelocateWithAlternative : public BinaryOperator
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateWithAlternative.h]: Maybe rename to `RelocateAlternative`.

File: tests/search/test_RelocateAlternative.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    # The old route costs 1544 + 1593 + 1475 = 4612. The proposed route
+    # C3 -> C2 costs 1476 + 828 + 2063 = 4367, for a delta of -245.
+    move = op.evaluate(sol.nodes[0], sol.nodes[3], cost_eval)
+    assert_equal(move, (-245, True))
+
+    op.apply(sol.nodes[0], sol.nodes[3])
+    assert_equal(str(sol.routes[0]), "C3 C2")
+
+
+def test_accounts_for_prizes_and_fixed_vehicle_costs():
```
Comment (N-Wouda) [inline tests/search/test_RelocateAlternative.py:74]: Can we also add a test where V is the start depot of an empty route (for that route's fixed cost component?)

File: pyvrp/cpp/search/RelocateWithAlternative.cpp (hunk 1 above)
Comment (leonlan) [reply pyvrp/cpp/search/RelocateWithAlternative.cpp]: I also changed this for ReplaceGroup

File: pyvrp/cpp/search/RelocateAlternative.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            costEvaluator.deltaCost(
+                deltaCost,
+                Route::Proposal(route->before(U->pos() - 1),
+                                route->between(U->pos() + 1, V->pos()),
+                                ClientSegment(data, client),
+                                route->after(V->pos() + 1)));
+
+            move_ = {deltaCost, alternative};
+            if (deltaCost < 0)
+                return;
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateAlternative.cpp]:
```suggestion
            
            if (deltaCost < 0)
            {
                move_ = {deltaCost, alternative};
                return;
            }
```

File: pyvrp/cpp/search/RelocateAlternative.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+void RelocateAlternative::evalWithinRoute(Route::Node *U,
+                                          Route::Node *V,
+                                          ClientGroup const &group,
+                                          CostEvaluator const &costEvaluator)
+{
+    auto const *route = U->route();
+    auto const &uData = data.client(U->idx());
+
+    if (U->pos() < V->pos())
+    {
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateAlternative.cpp]: These brackets (for if and else) are unnecessary

File: pyvrp/cpp/search/RelocateAlternative.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    auto const *vRoute = V->route();
+    auto const &uData = data.client(U->idx());
+    if (!uRoute || !vRoute || !uData.group || U == V || U == n(V))
+        return std::make_pair(0, false);
+
+    if (uRoute == vRoute && U->trip() != V->trip())
+        return std::make_pair(0, false);
+
+    auto const &group = data.group(*uData.group);
+    assert(group.mutuallyExclusive);
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateAlternative.cpp]: Do we still need this here? I think we can just get the group inside the later eval, rather than passing it as an argument.

File: pyvrp/cpp/search/RelocateAlternative.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    move_ = {};
+
+    auto const *uRoute = U->route();
+    auto const *vRoute = V->route();
+    auto const &uData = data.client(U->idx());
+    if (!uRoute || !vRoute || !uData.group || U == V || U == n(V))
+        return std::make_pair(0, false);
+
+    if (uRoute == vRoute && U->trip() != V->trip())
+        return std::make_pair(0, false);
```
Comment (N-Wouda) [inline pyvrp/cpp/search/RelocateAlternative.cpp]: This can move into `evalWithinRoute`, I think.

**Review by N-Wouda — APPROVED** [review]
LGTM if perf is also OK
